### DISCUSSION TITLE
Github Actions CI/CD pipeline (branch -> master -> release tag)

### DISCUSSION THREAD

**Discussion by tlfbrito** [community/community#26115] [Actions]
<p>HI there,</p>
<p>I have 3 workflow files:</p>
<p><strong>push.yml</strong> (Test &amp; lint when (people create a feature branch):</p>
<pre><code>on:
push:
branches:
-'\*/\*'  
-'\*'  
-'!master' 
</code></pre>
<p><strong>master.yml</strong> (To deploy to a Staging environment):</p>
<pre><code>on:
push:
branches:
- master# Push events on master branch
</code></pre>
<p>and finally I have <strong>release.yml</strong> (to deploy to production when we create a new tag):</p>
<pre><code>on:
push:
branches:
- releases/*# Push events to branches matching refs/heads/releases/\*
</code></pre>
<p>Unfortunately, when I publish a tag, all of the workflows are triggered. What I’m missing here?</p>
<p>Thanks for your help in advance.</p>

**Comment by stefanluptak** [comment]
<p>I am facing the same issue. It’s quite a show stopper, as we’re not able to migrate our CI/CD workflows from some rival platforms.</p>

**Comment by juulSme** [comment, marked as answer]
<p>9/6/2019: This is no longer relevant - it was a bug that has been fixed, see responses down below.</p>
<hr>
<p>I’ve had the same issue, it’s a pain in the ass. Github Actions is quite beta still, just wait until you want to get a service container working. Anyway, I’ve found the solution to your current problem. You need to exclude all tags from builds you want triggered on branch(es) only. For example:</p>
<pre><code>name: CI / Automated testing
on:
  push:
    tags:
      - '!refs/tags/*'
</code></pre>
<p>…triggers on all branches for my CI.</p>
<p>For my autodeploy-master-to-staging, I use this:</p>
<pre><code>name: Autodeploy
on:
  push:
    branches:
      - master
    tags:
      - '!refs/tags/*'
</code></pre>
<p>And finally, for my deploy-a-tagged-and-published-release-to-production, I use</p>
<pre><code>name: Production deploy
on:
  release:
    tags:
      - v*

jobs:
  prod-deploy:
    name: Production deploy
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        if: github.event.action == 'published'
        uses: actions/checkout@master
</code></pre>
<p>Most painfully, that little “if” in the final step has to be in EVERY step of the job, because there is no job-level if (yet).</p>
<p>I really like Github Actions, in general I like having the CI/CD nice and close to the code, but it is a far from mature solution at the moment.</p>

**Comment by jeremyepling** [comment]
<p>This is a bug. It should be fixed this week.</p>

**Comment by tlfbrito** [comment]
<p><a class="mention" href="/jeremyepling">@jeremyepling</a> Thanks! That is great!</p>
<p>Can you please clarify or point out the documentation on how to trigger the build when (and only when) a tag is published ?</p>

**Comment by jeremyepling** [comment]
<p>The fix is deployed. Now <code>tags</code> will only apply to pushed refs that match that tag. </p>
<p><a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#onpushpull_requesttagsbranches" rel="nofollow noopener">https://help.github.com/en/articles/workflow-syntax-for-github-actions#onpushpull_requesttagsbranches</a> is the right syntax. A documentation update with more details will go out today or tomorrow.</p>
<p>Here’s the updated line. “Defining only one of <code>tags:</code> or <code>branches:</code> will mean the workflow skips events affecting the other type of ref.”</p>

**Comment by chiefjester** [comment]
<p>They recently added events filter for webhook events so that you can have this instead:</p>
<pre><code>on:
  release:
      types: [published]
</code></pre>
<p>You can check the documentation <a href="https://help.github.com/en/articles/events-that-trigger-workflows#example-triggering-workflows-when-a-pull_request-is-assigned" title="triggering workflows when a pull request is assigned" rel="nofollow noopener">here.</a> That way, you don’t need to litter your workflow with if statements since they don’t have job-level if.</p>
<p>There’s also a setting for jobs to fail fast. If one job fails, it fails everything and doesn’t need the rest of jobs to finish here’s a link of their <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast" rel="nofollow noopener">documentation</a>.</p>

**Comment by juulSme** [comment]
<p>It has been deployed indeed :slight_smile: my old solution is no longer required. To build for all branches and not for tags, I now use:</p>
<pre><code>name: CI / Automated testing
on:
  push:
    branches:

jobs:
  ...
</code></pre>

**Comment by juulSme** [comment]
<p>Thanks for that, I’m adapting to it now. I have to say, things are improving rapidly :slight_smile:</p>
